engine/common/follower/engine.go

```diff
@@ -209,13 +217,44 @@ func (e *Engine) processQueuedBlocks(doneSignal <-chan struct{}) error {
 		msg, ok := e.pendingBlocks.Pop()
 		if ok {
 			batch := msg.(flow.Slashable[[]*messages.BlockProposal])
+			if len(batch.Message) < 1 {
+				continue
+			}
+			blocks := make([]*flow.Block, 0, len(batch.Message))
 			for _, block := range batch.Message {
-				err := e.processBlockProposal(batch.OriginID, block)
-				if err != nil {
-					return fmt.Errorf("could not handle block proposal: %w", err)
+				blocks = append(blocks, block.Block.ToInternal())
+			}
+
+			firstBlock := blocks[0].Header
+			lastBlock := blocks[len(blocks)-1].Header
+			log := e.log.With().
+				Hex("origin_id", batch.OriginID[:]).
+				Str("chain_id", lastBlock.ChainID.String()).
+				Uint64("first_block_height", firstBlock.Height).
+				Uint64("first_block_view", firstBlock.View).
+				Uint64("last_block_height", lastBlock.Height).
+				Uint64("last_block_view", lastBlock.View).
+				Int("range_length", len(blocks)).
+				Logger()
+
+			latestFinalizedView := e.finalizedBlockTracker.NewestBlock().View
+			submitConnectedBatch := func(blocks []*flow.Block) {
+				e.submitConnectedBatch(log, latestFinalizedView, batch.OriginID, blocks)
+			}
+
+			// extract sequences of connected blocks and schedule them for further processing
+			parentID := blocks[0].ID()
+			indexOfLastConnected := 0
+			for i := 1; i < len(blocks); i++ {
+				if blocks[i].Header.ParentID != parentID {
+					submitConnectedBatch(blocks[indexOfLastConnected:i])
+					indexOfLastConnected = i
 				}
-				e.engMetrics.MessageHandled(metrics.EngineFollower, metrics.MessageBlockProposal)
+				parentID = blocks[i].Header.ID()
 			}
+			submitConnectedBatch(blocks[indexOfLastConnected:])
+
+			e.engMetrics.MessageHandled(metrics.EngineFollower, metrics.MessageBlockProposal)
```

Comment:
This could be either a proposal, or a synced list of blocks. But we're always assuming it's a proposal for the metrics.

I'm not sure the best thing to do here. We could not bother differentiating between the message types, or pass in the message type somehow here.

Reply:
As I have added a separate queue for proposals received from compliance interface we now differentiate between those metrics.